yente/search/base.py

```diff
@@ -74,3 +91,317 @@ async def close_es() -> None:
+                },
+            )
+        finally:
+            await self._remove_write_block(index)
+
+    async def delete_index(self, index: str) -> None:
+        """
+        Delete a given index if it exists.
+        """
+        try:
+            await self.client.indices.delete(index=index)
+        except NotFoundError:
+            pass
+
+    async def get_backing_indexes(self, name: str) -> List[str]:
+        resp = await self.client.indices.get_alias(name=name)
+        return list(resp.keys())
+
+    async def index_exists(self, index: str) -> bool:
+        exists = await self.client.indices.exists(index=index)
+        if exists.body:
+            log.info("Index is up to date.", index=index)
+            return True
+        return False
+
+    async def rollover(self, alias: str, new_index: str, prefix: str = "") -> None:
+        """
+        Remove all existing indices with a given prefix from the alias and add the new one.
+        """
+        actions = []
+        actions.append({"remove": {"index": f"{prefix}*", "alias": alias}})
+        actions.append({"add": {"index": new_index, "alias": alias}})
+        await self.client.indices.update_aliases(actions=actions)
+        return None
+
+    async def count(self, index: str) -> int:
+        resp = await self.client.count(index=index)
+        return int(resp["count"])
+
+    async def get_alias_sources(self, alias: str) -> List[str]:
+        resp = await self.client.indices.get_alias(name=alias)
+        return list(resp.keys())
+
+    async def refresh(self, index: str) -> None:
+        await self.client.indices.refresh(index=index)
+
+    async def add_alias(self, index: str, alias: str) -> None:
+        await self.client.indices.put_alias(index=index, name=alias)
+
+    async def update(
+        self, entities: AsyncGenerator[Dict[str, Any], None], index_name: str
+    ) -> Tuple[int, int]:
+        """
+        Update the index with the given entities in bulk.
+        Return a tuple of the number of successful and failed operations.
+        """
+        resp = await async_bulk(
+            client=self.client,
+            actions=self._entity_iterator(entities, index_name),
+            chunk_size=500,
+            raise_on_error=True,
+            stats_only=True,
+        )
+        return cast(Tuple[int, int], resp)
+
+    async def _entity_iterator(
+        self, async_entities: AsyncGenerator[Dict[str, Any], None], index: str
+    ) -> AsyncGenerator[Dict[str, Any], Any]:
+        async for data in async_entities:
+            yield self._to_operation(data, index)
+
+    def _delete_operation(self, doc_id: str, index: str) -> Dict[str, Any]:
+        return {
+            "_op_type": "delete",
+            "_index": index,
+            "_id": doc_id,
+        }
+
+    def _update_operation(
+        self, entity: Dict[str, Any], doc_id: str, index: str
+    ) -> Dict[str, Any]:
+        return {
+            "_op_type": "update",
```

Comment:
Are we sure we want to update and not index here?

Reply:
We can use index as well as long as we're sure we'll never ship partial entities. Then we can combine _create and _update into _upsert.